pallets/corporate-actions/src/ballot/benchmarking.rs

```diff
@@ -0,0 +1,125 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg(feature = "runtime-benchmarks")]
+
+use super::*;
+use crate::benchmarking::{set_ca_targets, setup_ca};
+use core::iter;
+use frame_benchmarking::benchmarks;
+use pallet_timestamp::Module as Timestamp;
+use polymesh_common_utilities::benchs::User;
+
+const MAX_CHOICES: u32 = 1000;
+const MAX_TARGETS: u32 = 1000;
+
+const RANGE: BallotTimeRange = BallotTimeRange {
+    start: 3000,
+    end: 4000,
+};
+
+fn meta(n_motions: u32, n_choices: u32) -> BallotMeta {
+    let motion = Motion {
+        title: "".into(),
+        info_link: "".into(),
+        choices: iter::repeat("".into()).take(n_choices as usize).collect(),
+    };
+    let motions = iter::repeat(motion).take(n_motions as usize).collect();
+    BallotMeta {
+        title: "".into(),
+        motions,
+    }
+}
+
+fn attach<T: Trait>(n_motions: u32, n_choices: u32) -> (User<T>, CAId) {
+    let meta = meta(n_motions, n_choices);
+    let (owner, ca_id) = setup_ca::<T>(CAKind::IssuerNotice);
+    <Module<T>>::attach_ballot(owner.origin().into(), ca_id, RANGE, meta, true).unwrap();
+    (owner, ca_id)
+}
+
+benchmarks! {
+    _ {}
+
+    attach_ballot {
+        let j in 0..MAX_CHOICES;
+
+        let meta = meta(1, j);
+        let (owner, ca_id) = setup_ca::<T>(CAKind::IssuerNotice);
+    }: _(owner.origin(), ca_id, RANGE, meta, true)
+    verify {
+        ensure!(TimeRanges::get(ca_id) == Some(RANGE), "ballot not created");
+    }
+
+    vote {
+        let j in 0..MAX_CHOICES;
+        let k in 0..MAX_TARGETS;
+
+        // Attach and prepare to vote.
+        let (owner, ca_id) = attach::<T>(1, j);
+        #[cfg(feature = "std")]
+        <Timestamp<T>>::set_timestamp(3000.into());
+
+        // Change targets, as they are read in voting.
+        set_ca_targets::<T>(ca_id, k);
+
+        // Construct the voting list.
+        let votes = (0..j)
+            .map(|j| BallotVote {
+                power: 0.into(),
+                fallback: (j as u16).checked_sub(1),
+            })
+            .collect::<Vec<_>>();
+
+        // Vote already to force a longer code path.
+        <Module<T>>::vote(owner.origin().into(), ca_id, votes.clone()).unwrap();
+        let results = votes.iter().map(|v| v.power).collect::<Vec<_>>();
+    }: _(owner.origin(), ca_id, votes)
+    verify {
+        ensure!(<Results<T>>::get(ca_id) == results, "voting results are wrong")
+    }
+
+    change_end {
+        let (owner, ca_id) = attach::<T>(0, 0);
+    }: _(owner.origin(), ca_id, 5000)
+    verify {
+        ensure!(TimeRanges::get(ca_id).unwrap().end == 5000, "range not changed");
+    }
+
+    change_meta {
+        let j in 0..MAX_CHOICES;
+
+        let (owner, ca_id) = attach::<T>(0, 0);
+        let meta = meta(1, j);
+        let meta2 = meta.clone();
+    }: _(owner.origin(), ca_id, meta)
+    verify {
+        ensure!(Metas::get(ca_id).unwrap() == meta2, "meta not changed");
```

Comment:
```suggestion
    }: _(owner.origin(), ca_id, meta.clone())
    verify {
        ensure!(Metas::get(ca_id).unwrap() == meta, "meta not changed");
```

Reply:
This is intentional. The macro will expand everything inside `_(owner.origin(), ca_id, meta)` into an `$eval:block` fragment and make that the actual logic being benchmarked. If the `.clone()` is moved into the `_(...)` then you are also adding the cost of the clone into what is being benchmarked, thereby making things less accurate (tiny difference I know). By having `let meta2` outside, the overhead is at most one memcpy, which is less.

Reply:
Actually the cost of the clone should not be added to the benchmark, because it is not a direct call. The macro expands to  a creation of `T::Call::...` object, which contains the arguments. The creation of that object is done before the call to the specific `dispatch` function where benchmark is executed.

Reply:
My reading of the macro is that we have the following transformation:
```rust
$name:ident { $( $code:tt )* }: _ ( $origin:expr $( , $arg:expr )* )
```
==>
```rust
			$name { $( $code )* }: {
				<
					Call<T $(, $instance)? > as $crate::frame_support::traits::UnfilteredDispatchable
				>::dispatch_bypass_filter(
					Call::<T $(, $instance)? >::$dispatch($($arg),*), $origin.into()
				)?;
			}
```
The block to the right there is the `$eval:block`. This gets plumbed down right into the benchmark itself, where we have:
```rust
			fn instance(
				&self,
				components: &[($crate::BenchmarkParameter, u32)],
				verify: bool
			) -> Result<Box<dyn FnOnce() -> Result<(), &'static str>>, &'static str> {
				// Other stuff...

				Ok(Box::new(move || -> Result<(), &'static str> {
					$eval;
					if verify {
						$postcode;
					}
					Ok(())
				}))
			}
```
So the actual benchmark being executed is really this boxed closure, no?

Reply:
It is going to be something like:
```
Ok(Box::new(move || -> Result<(), &'static str> {
   {
       <Call<T> as frame_benchmarking::frame_support::traits::UnfilteredDispachable>::dispatch_bypass_filter(
            Call::<T>::change_meta( ca_id, meta.clone()),
            owner.origin()
        )?;
       }
       if verify {
        ....
       }
```

The object `Call::<T>::change_meta(...)` is created completely before dispatch it, so the `clone` fn is not going to be part of the bench.

Reply:
After a deeper look, as @Centril said, the creation of the `Call` instance is *inside* the bench.